NitroxClient/GameLogic/Spawning/Metadata/Extractor/GenericEntityMetadataExtractor.cs

```diff
@@ -0,0 +1,63 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using NitroxModel.DataStructures.GameLogic.Entities.Metadata;
+using NitroxModel.DataStructures.Util;
+using UnityEngine;
+
+namespace NitroxClient.GameLogic.Spawning.Metadata.Extractor;
+
+public abstract class GenericEntityMetadataExtractor<I, O> : EntityMetadataExtractor where O : EntityMetadata
+{
+    public abstract O Extract(I entity);
+
+    public override Optional<EntityMetadata> From(object o)
+    {
+        EntityMetadata result = Extract((I)o);
+
+        return Optional.Of(result);
+    }
+}
+
+public abstract class EntityMetadataExtractor
+{
+    public abstract Optional<EntityMetadata> From(object o);
+
+    private static readonly Dictionary<Type, EntityMetadataExtractor> processors;
+
+    static EntityMetadataExtractor()
+    {
+        processors = Assembly.GetExecutingAssembly()
+                             .GetTypes()
+                             .Where(t => typeof(EntityMetadataExtractor).IsAssignableFrom(t) &&
+                                    t.IsClass && !t.IsAbstract)
+                             .Select(Activator.CreateInstance)
+                             .Cast<EntityMetadataExtractor>()
+                             .ToDictionary(p => p.GetType().BaseType.GetGenericArguments()[0]);    
+    }
+
+    public static Optional<EntityMetadata> Extract(object o)
+    {
+        if (processors.TryGetValue(o.GetType(), out EntityMetadataExtractor extractor))
+        {
+            return extractor.From(o);
+        }
+
+        return Optional.Empty;
+    }
+
+    public static Optional<EntityMetadata> Extract(GameObject o)
```

Comment:
That could be problematic with multiple viable Components on the GameObject. I would favor using a Component as the parameter.

Reply:
Yes, this is one issue that I currently see.  However, the only codepath that invokes this is when we drop an item.  In that case, we don't actually know which component will source metadata (if any).  I will explore ways to improve this in a follow-up when implementing the seaglide.  Potentially we can use the techtype to prioritize specific components or we can define a global priority.  I'm not sure the best approach until implementing a few more and seeing what works best.